File: 39-combination-sum/39-combination-sum.cpp
class Solution {
    // Time : 2^k * k.
    // Space: k * x
public:
    void findCombination(int ind, int target, vector<int> &arr, vector<vector<int>>&ans, vector<int> &ds)
    {
        //base case:
        if(ind == arr.size()){
            if(target == 0){
                ans.push_back(ds);
            }
            return;
        }
        //pick up the element:
        if(arr[ind] <= target)
        {
            ds.push_back(arr[ind]);
            findCombination(ind, target-arr[ind], arr, ans, ds);
            
            ds.pop_back();
        }
        //not pick an element;
        findCombination(ind+1, target, arr, ans, ds);
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        // creating vector of vector to store our ans:
        vector<vector<int>> ans;
        //creating a vector to store a particular combination:
        vector<int> ds;
        findCombination(0,target,candidates, ans, ds);
        
        return ans;
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public
        void findCombination(int ind,int []arr, int target, List<Integer>ds,List<List<Integer>> ans){
        // base case:
        if(ind == arr.length){
            if(target == 0)
                ans.add(new ArrayList<>(ds));
            return;
        }
        // pick case:
        if(arr[ind]<=target){
            ds.add(arr[ind]);
            findCombination(ind,arr,target-arr[ind],ds,ans);
            ds.remove(ds.size()-1);
        }
        // not pick case:
        findCombination(ind+1,arr,target,ds,ans);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();
        
        findCombination(0,candidates,target,ds,ans);
        return ans;
    }
}